File: hooks/stop-scan-fullwidth-punctuation.py
#!/usr/bin/env python3
"""Stop hook: scan all files under cwd for fullwidth punctuation."""

import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from git_utils import git_tracked_files

_FULLWIDTH_PUNCTUATION = re.compile(
    r'[。\、，；：！？「」『』【】〈〉《》〔〕（）……——·〜～"'']'
)
_NOQA_MARKER = "# noqa: fullwidth-punctuation"


def _check_file(path: Path) -> list[str]:
    """Return list of lines with fullwidth punctuation."""
    try:
        text = path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return []

    violations: list[str] = []
    for lineno, line in enumerate(text.splitlines(), 1):
        if _NOQA_MARKER in line:
            continue
        if _FULLWIDTH_PUNCTUATION.search(line):
            found = _FULLWIDTH_PUNCTUATION.findall(line)
            violations.append(f"{path}:{lineno} {repr(found)}")
    return violations


def main() -> None:
    data = json.load(sys.stdin)

    if data.get("stop_hook_active"):
        return

    if data.get("permission_mode") == "plan":
        return

    cwd = data.get("cwd", "")
    if not cwd:
        return

    all_violations: list[str] = []
    for file_path in sorted(git_tracked_files(Path(cwd))):
        if not file_path.is_file():
            continue

        all_violations.extend(_check_file(file_path))

    if all_violations:
        detail = "\n".join(all_violations[:50])
        if len(all_violations) > 50:
            detail += f"\n  ... 他 {len(all_violations) - 50} 件"
        json.dump(
            {
                "decision": "block",
                "reason": (
                    f"全角句読点が {len(all_violations)} 箇所で見つかりました:\n"
                    f"{detail}\n\n"
                    f"半角記号を使用してください。"
                    f"除外する場合は行末に {_NOQA_MARKER} を追加してください。"
                ),
            },
            sys.stdout,
        )


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/stop_scan_fullwidth_punctuation.rs
use regex::Regex;

use crate::git::git_tracked_files;
use crate::input::HookInput;
use crate::output::block;

const NOQA_MARKER: &str = "# noqa: fullwidth-punctuation";

pub fn run(input: &HookInput) {
    if input.stop_hook_active {
        return;
    }
    if input.permission_mode == "plan" {
        return;
    }
    let cwd = &input.cwd;
    if cwd.is_empty() {
        return;
    }

    let re = Regex::new(
        r"[。，；：！？「」『』【】〈〉《》〔〕（）……——·〜～\x{201c}\x{201d}\x{2018}\x{2019}]",
    )
    .unwrap();
    let root = std::path::Path::new(cwd);
    let mut all_violations: Vec<String> = Vec::new();

    for file_path in git_tracked_files(root, &[]) {
        if !file_path.is_file() {
            continue;
        }
        let text = match std::fs::read_to_string(&file_path) {
            Ok(t) => t,
            Err(_) => continue,
        };
        for (i, line) in text.lines().enumerate() {
            if line.contains(NOQA_MARKER) {
                continue;
            }
            if re.is_match(line) {
                let found: Vec<&str> = re.find_iter(line).map(|m| m.as_str()).collect();
                all_violations.push(format!("{}:{} {:?}", file_path.display(), i + 1, found));
            }
        }
    }

    if !all_violations.is_empty() {
        let detail = all_violations
            .iter()
            .take(50)
            .cloned()
            .collect::<Vec<_>>()
            .join("\n");
        let extra = if all_violations.len() > 50 {
            format!("\n  ... 他 {} 件", all_violations.len() - 50)
        } else {
            String::new()
        };
        block(&format!(
            "全角句読点が {} 箇所で見つかりました:\n{detail}{extra}\n\n半角記号を使用してください。除外する場合は行末に {NOQA_MARKER} を追加してください。",
            all_violations.len()
        ));
    }
}
